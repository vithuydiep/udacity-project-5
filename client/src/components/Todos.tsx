import dateFormat from 'dateformat'
import { History } from 'history'
import update from 'immutability-helper'
import * as React from 'react'
import {
  Button,
  Checkbox,
  Divider,
  Grid,
  Header,
  Icon,
  Input,
  Image,
  Loader,
  Confirm
} from 'semantic-ui-react'

import { createTodo, deleteTodo, getSearchTodo, getTodos, patchTodo } from '../api/todos-api'
import Auth from '../auth/Auth'
import { Todo } from '../types/Todo'

interface TodosProps {
  auth: Auth
  history: History
}

interface TodosState {
  todos: Todo[]
  newTodoName: string
  loadingTodos: boolean
  keyword: string
  openModel: boolean
  todoId: string
}

export class Todos extends React.PureComponent<TodosProps, TodosState> {
  state: TodosState = {
    todos: [],
    newTodoName: '',
    loadingTodos: true,
    keyword: '',
    openModel: false,
    todoId: ''
  }

  handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ newTodoName: event.target.value })
  }

  handleKeyWordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ keyword: event.target.value })
  }

  onEditButtonClick = (todoId: string) => {
    this.props.history.push(`/todos/${todoId}/edit`)
  }

  onLoadingData = () => {
    this.setState({
      ...this.state,
      loadingTodos: true
    })
  }

  onTodoCreate = async (event: React.ChangeEvent<HTMLButtonElement>) => {
    try {
      const dueDate = this.calculateDueDate()
      const newTodo = await createTodo(this.props.auth.getIdToken(), {
        name: this.state.newTodoName,
        dueDate
      })
      this.setState({
        todos: [...this.state.todos, newTodo],
        newTodoName: ''
      })
    } catch {
      alert('Todo creation failed')
    }
  }

  onSearchTodo = async (event: React.ChangeEvent<HTMLButtonElement>) => {
    this.onLoadingData()
    try {
      if (this.state.keyword) {
        const todos = await getSearchTodo(this.props.auth.getIdToken(), this.state.keyword)
        this.setState({
          todos,
          loadingTodos: false
        })
      } else {
        const todos = await getTodos(this.props.auth.getIdToken())
        this.setState({
          todos,
          loadingTodos: false
        })
      }

    } catch {
      alert('Todo creation failed')
    }
  }


  onTodoDelete = async (todoId: string) => {
    this.setState({
      ...this.state,
      todoId,
      openModel: true
    })
  }

  onTodoCheck = async (pos: number) => {
    try {
      const todo = this.state.todos[pos]
      await patchTodo(this.props.auth.getIdToken(), todo.todoId, {
        name: todo.name,
        dueDate: todo.dueDate,
        done: !todo.done
      })
      this.setState({
        todos: update(this.state.todos, {
          [pos]: { done: { $set: !todo.done } }
        })
      })
    } catch {
      alert('Todo deletion failed')
    }
  }

  onHandleCancel = () => {
    this.setState({
      ...this.state,
      openModel: false
    })
  }

  onHandleConfirm = async () => {
    try {
      await deleteTodo(this.props.auth.getIdToken(), this.state.todoId)
      this.setState({
        todos: this.state.todos.filter(todo => todo.todoId !== this.state.todoId),
        openModel: false
      })
    } catch {
      alert('Todo deletion failed')
    }
  }

  async componentDidMount() {
    try {
      const todos = await getTodos(this.props.auth.getIdToken())
      this.setState({
        todos,
        loadingTodos: false
      })
    } catch (e) {
      alert(`Failed to fetch todos: ${(e as Error).message}`)
    }
  }

  render() {
    return (
      <div>
        <Header as="h1">TODOs</Header>

        <Confirm
          content='Are you sure you want to delete this task?'
          open={this.state.openModel}
          onCancel={this.onHandleCancel}
          onConfirm={this.onHandleConfirm}
        />

        {this.renderCreateTodoInput()}

        {this.renderSearchTodoInput()}

        {this.renderTodos()}
      </div>
    )
  }

  renderCreateTodoInput() {
    return (
      <Grid.Row>
        <Grid.Column width={16}>
          <Input
            action={{
              color: 'teal',
              labelPosition: 'left',
              icon: 'add',
              content: 'New task',
              onClick: this.onTodoCreate
            }}
            fluid
            actionPosition="left"
            placeholder="To change the world..."
            onChange={this.handleNameChange}
          />
        </Grid.Column>
        <Grid.Column width={16}>
          <Divider />
        </Grid.Column>
      </Grid.Row>
    )
  }

  renderSearchTodoInput() {
    return (
      <Grid.Row>
        <Grid.Column width={16}>
          <Input
            action={{
              color: 'teal',
              labelPosition: 'left',
              icon: 'search',
              content: 'Search task',
              onClick: this.onSearchTodo
            }}
            fluid
            actionPosition="left"
            placeholder="Press keyword to search..."
            onChange={this.handleKeyWordChange}
          />
        </Grid.Column>
        <Grid.Column width={16}>
          <Divider />
        </Grid.Column>
      </Grid.Row>
    )
  }

  renderTodos() {
    if (this.state.loadingTodos) {
      return this.renderLoading()
    }

    return this.renderTodosList()
  }

  renderLoading() {
    return (
      <Grid.Row>
        <Loader indeterminate active inline="centered">
          Loading TODOs
        </Loader>
      </Grid.Row>
    )
  }

  renderTodosList() {
    return (
      <Grid padded>
        {this.state.todos.map((todo, pos) => {
          return (
            <>
              <Grid.Row key={todo.todoId} color={!todo.done ? 'grey' : undefined} >
                <Grid.Column width={1} verticalAlign="middle">
                  <Checkbox
                    onChange={() => this.onTodoCheck(pos)}
                    checked={todo.done}
                  />
                </Grid.Column>
                <Grid.Column width={10} verticalAlign="middle">
                  {todo.name}
                </Grid.Column>
                <Grid.Column width={3} floated="right">
                  {todo.dueDate}
                </Grid.Column>
                <Grid.Column width={1} floated="right">
                  <Button
                    icon
                    color="blue"
                    onClick={() => this.onEditButtonClick(todo.todoId)}
                  >
                    <Icon name="pencil" />
                  </Button>
                </Grid.Column>
                <Grid.Column width={1} floated="right">
                  <Button
                    icon
                    color="red"
                    onClick={() => this.onTodoDelete(todo.todoId)}
                  >
                    <Icon name="delete" />
                  </Button>
                </Grid.Column>
                {todo.attachmentUrl && (
                  <Image src={todo.attachmentUrl} size="small" wrapped onError={(i: any) => i.target.style.display = 'none'} />
                )}
              </Grid.Row>
              <Grid.Column width={16}>
                <Divider />
              </Grid.Column>
            </>
          )
        })}
      </Grid>
    )
  }

  calculateDueDate(): string {
    const date = new Date()
    date.setDate(date.getDate() + 7)

    return dateFormat(date, 'yyyy-mm-dd') as string
  }
}
