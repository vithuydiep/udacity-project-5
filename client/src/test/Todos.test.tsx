import { act, render, waitFor,screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import Auth from '../auth/Auth'
import { Todos } from '../components/Todos'
import { History } from 'history'
import * as API from '../api/todos-api'

const auth: Auth = {
  accessToken: 'fff',
  idToken: 'ffff',
  expiresAt: 'ffff',
  auth0: jest.fn(),
  getAccessToken: jest.fn(),
  getIdToken: jest.fn(),
  handleAuthentication: jest.fn(),
  history: jest.fn(),
  isAuthenticated: jest.fn(),
  login: jest.fn(),
  logout: jest.fn(),
  renewSession: jest.fn(),
  setSession: jest.fn()
}
const historyMock: History = {
  length: 2,
  action: 'REPLACE',
  location: {
    pathname: '/',
    search: '',
    hash: '',
    key: 'se921n',
    state: ''
  },
  block: jest.fn(),
  createHref: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  listen: jest.fn(),
  push: jest.fn(),
  replace: jest.fn()
}
const mockData = [
  {
    todoId: '4a093a6f-3927-4a22-9754-085b6e961eca',
    userId: 'google-oauth2|103303390409976191104',
    attachmentUrl:
      'https://my-serverless-s3-dev.s3.amazonaws.com/4a093a6f-3927-4a22-9754-085b6e961eca',
    dueDate: '2023-05-09',
    createdAt: '2023-05-02T06:37:18.063Z',
    name: 'learning',
    done: true
  }
]
const TodoListComponent = <Todos auth={auth} history={historyMock} />

describe('check render todo list', async () => {
  beforeEach(async () => {
    jest
      .spyOn(API, 'getTodos')
      .mockImplementation(() => Promise.resolve(mockData))
    await act(async () => {
      render(TodoListComponent)
    })
  })

  it('should render and match the snapshot', async () => {
    const { asFragment } = render(TodoListComponent)
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it('should render fetch api', () => {
    expect(API.getTodos).toBeCalled()
  })

  it('should show correct data', () => {
    const testLabel = screen.getByText('learning')
    expect(testLabel).toBeInTheDocument()
  })

  it('should call to open modal when click on delete button', async () => {
    const btnDelete = screen.getAllByRole('button')
    fireEvent.click(btnDelete[3])
    expect(screen.getByText('Are you sure you want to delete this task?')).toBeInTheDocument()
  })
})
