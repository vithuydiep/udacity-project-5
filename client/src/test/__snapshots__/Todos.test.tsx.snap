// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check render todo list should render and match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="ui header"
    >
      TODOs
    </h1>
    <div
      class="row"
    >
      <div
        class="sixteen wide column"
      >
        <div
          class="ui fluid left action input"
        >
          <button
            class="ui teal icon left labeled button"
          >
            <i
              aria-hidden="true"
              class="add icon"
            />
            New task
          </button>
          <input
            placeholder="To change the world..."
            type="text"
          />
        </div>
      </div>
      <div
        class="sixteen wide column"
      >
        <div
          class="ui divider"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="sixteen wide column"
      >
        <div
          class="ui fluid left action input"
        >
          <button
            class="ui teal icon left labeled button"
          >
            <i
              aria-hidden="true"
              class="search icon"
            />
            Search task
          </button>
          <input
            placeholder="Press keyword to search..."
            type="text"
          />
        </div>
      </div>
      <div
        class="sixteen wide column"
      >
        <div
          class="ui divider"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="ui active indeterminate text centered inline loader"
      >
        Loading TODOs
      </div>
    </div>
  </div>
</DocumentFragment>
`;
